### PR TITLE
Implement DisplayString stuff

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/messagestream/hotkey.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#ifndef GAME_DLL
+HotKeyManager *g_theHotKeyManager;
+#endif
+
+Utf8String HotKeyManager::Search_Hot_Key(const Utf16String &key)
+{
+    if (!key.Is_Empty()) {
+        for (const unichar_t *i = key.Str(); i != nullptr && *i != '\0'; i++) {
+            if (*i == '&') {
+                Utf16String str(Utf16String::s_emptyString);
```
Comment (xezon) [inline src/game/client/messagestream/hotkey.cpp:27]: Constructing with `Utf16String::s_emptyString` for sure?

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#include "w3ddisplaystringmanager.h"
+#include "drawgroupinfo.h"
+#include "gameclient.h"
+#include "gametext.h"
+#include "globallanguage.h"
+#include "w3ddisplaystring.h"
+
+W3DDisplayStringManager::W3DDisplayStringManager()
+{
+    for (int i = 0; i < 10; i++) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: `i < ARRAY_SIZE(m_groupNumeralStrings)`, several times in this class.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            m_currentCheckpoint = nullptr;
+        }
+
+        string->Delete_Instance();
+    }
+}
+
+DisplayString *W3DDisplayStringManager::Get_Group_Numeral_String(int num)
+{
+    if (num > 9) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: `num > ARRAY_SIZE(m_groupNumeralStrings) - 1`

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+void W3DDisplayStringManager::Update()
+{
+    W3DDisplayString *str = static_cast<W3DDisplayString *>(m_stringList);
+
+    if (m_currentCheckpoint != nullptr) {
+        str = static_cast<W3DDisplayString *>(m_currentCheckpoint);
+    }
+
+    uint32_t frame = g_theGameClient->Get_Frame();
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: curr_frame

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    int i = 9;
+
+    do {
+        if (str == nullptr) {
+            break;
+        }
+
+        unsigned int frame2 = str->m_frame;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: draw_frame

File: src/platform/w3dengine/client/w3ddisplaystring.cpp [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    FontCharsClass *font = m_sentence.m_font;
+    int width = 0;
+    int i = 0;
+
+    if (font != nullptr) {
+        const unichar_t *str = m_textString.Str();
+        unichar_t c = *str;
+
+        for (const unichar_t *j = str + 1; c != '\0'; j++) {
+            if (char_count != 0xFFFFFFFF && i >= char_count) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystring.cpp]: -1

File: src/platform/w3dengine/client/w3ddisplaystring.cpp [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+{
+    FontCharsClass *font = m_sentence.m_font;
+    int width = 0;
+    int i = 0;
+
+    if (font != nullptr) {
+        const unichar_t *str = m_textString.Str();
+        unichar_t c = *str;
+
+        for (const unichar_t *j = str + 1; c != '\0'; j++) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystring.cpp]: next

File: src/platform/w3dengine/client/w3ddisplaystring.cpp [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+{
+    m_hotKeyState = state;
+    m_hotKeyColor = color;
+    m_sentence.m_processAmpersand = true;
+    Notify_Text_Changed();
+}
+
+void W3DDisplayString::Set_Clip_Region(IRegion2D *region)
+{
+    int *i = &m_clipRegion.lo.x;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystring.cpp]: `i` seems unused. Can be removed.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+void W3DDisplayStringManager::Update()
+{
+    W3DDisplayString *str = static_cast<W3DDisplayString *>(m_stringList);
+
+    if (m_currentCheckpoint != nullptr) {
+        str = static_cast<W3DDisplayString *>(m_currentCheckpoint);
+    }
+
+    uint32_t frame = g_theGameClient->Get_Frame();
+    int i = 9;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: `ARRAY_SIZE(m_groupNumeralStrings) - 1`

File: src/game/client/messagestream/hotkey.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/messagestream/hotkey.cpp:27]: Yes.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp (hunk 5 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystring.cpp (hunk 6 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystring.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystring.cpp (hunk 7 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystring.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystring.cpp (hunk 8 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystring.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp (hunk 9 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: fixed.

File: src/platform/w3dengine/client/w3ddisplaystringmanager.cpp (hunk 3 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3ddisplaystringmanager.cpp]: fixed.

**Review by xezon — APPROVED** [review]
Strings look normal. Hotkeys look normal and work.

**Pull request merged**